Use a "newtype", and add relative humidity measurement

tim-seoss opened the pull request:
Any comments welcome, see also #2.  I've not tested this on an embedded Linux device (but it compiles! :-) ).

I have tested it on an stm32f103 tho'.

LMK what you think.

tim-seoss committed:
Reflect crate name.

tim-seoss committed:
Use newtype pattern, like std::time::Duration::as_secs etc.

dbrgn commented on src/lib.rs:
```diff
@@ -24,6 +24,20 @@ pub enum Error<E> {
 pub const ADDRESS: u8 = 0x40;
 
+pub struct MilliCelsius(i32);
+pub struct RHPerCentMille(i32);
```
I'd simply call those types `Temperature` and `RelativeHumidity`. The meaning of the internal number is already conveyed by the method names.

tim-seoss replied:
When I wrote that, I was going by the Rust Style Guidelines, which has a closely analogous example, and names the types according to the units used:

https://doc.rust-lang.org/1.0.0/style/features/types/newtype.html

I suppose I feel that:
```rust
fn convert_raw_temperature(temp_raw: u16) -> MilliCelsius {
```
is clearer than:
```rust
fn convert_raw_temperature(temp_raw: u16) -> Temperature {
```
... especially as the formula quoted from the datasheet (in the comments) calculates its answer in Celsius, but the fixed point implementation in the function itself calculates milli-Celsius instead, but this isn't obvious at first glance...

Any thoughts?

dbrgn replied:
The internal implementation does not really matter, as long as it represents a temperature and as long as it's clear what the values that are returned by the methods mean.

dbrgn commented on src/lib.rs:
```diff
@@ -113,6 +142,14 @@ where
+///
+/// Formula (datasheet 6.1): -6 + 125 * (val / 2^16)
+/// The implementation is equivalent for fixed point (3 decimal places).
+fn raw_to_per_cent_mille_rh(humidity_raw: u16) -> RHPerCentMille {
```
The name of this function is inconsistent with `convert_temperature`. I'd either use `convert_humidity` here, or rename both functions to `convert_raw_temperatuer` / `convert_raw_humidity`.

tim-seoss replied:
done

dbrgn reviewed (changes requested):
Thanks!

tim-seoss committed:
Add RH read and example (to match temperature functionality).

tim-seoss committed:
Zero out status bits (bit 0 is unassigned but not guarenteed to read 0).

tim-seoss committed:
Make register -> units conversion fn names consistent.

tim-seoss committed:
Qualify hal usage (by cargo "fix --edition").

tim-seoss committed:
Update to Rust edition 2018

tim-seoss commented:
I've also added a change to zero out the status bits.  These are currently returning zero on the sht21, but the datasheet doesn't guarantee this for bit zero, so I thought it was best to zero them to be on the safe side and to be consistent with the humidity conversion.  Also updated to 2018 edition.

tim-seoss commented:
Any further thoughts?  Was thinking of tackling a couple of things in #3 but it'd be good to get this resolved first if poss...

Cheers,

Tim.

dbrgn commented on src/lib.rs:
```diff
@@ -24,6 +24,20 @@ pub enum Error<E> {
+}
+
+impl RHPerCentMille {
+    pub fn as_percentmille(&self) -> i32 { self.0 }
```
Is `percentmille` a real word? It basically means "one thousandth of one hundredth".

It seems that `as_permille` would be the proper name: https://en.wikipedia.org/wiki/Per_mille

linuxtim replied:
Thanks for your comments - now it's my turn to be a bit (ahem) slow to reply...

> I'd prefer newtypes that are named after their meaning, not after their internal units

OK, I'll update the PR accordingly.

> It seems that `as_permille` would be the proper name: https://en.wikipedia.org/wiki/Per_mille

My reading of Wikipedia, is:

| Name | Denominator |
| ---- | -------- |
| Percent (%) | 10² |
| Per mille (‰) | 10³ |
| Per cent mille (pcm) | 10⁵ |
| Parts per million (ppm) | 10⁶ |

https://en.wikipedia.org/wiki/Per_cent_mille

... and the code currently calculates integers which are:
- Temperature in units of thousandths of a degree Celsius
- Relative Humidity in units of thousandths of a percent RH (pcm)

... so I think pcm is the correct name for the return type of the optimised integer RH calc at the moment.

Alternatively It might be less weird to alter the function to calculate ppm instead (since ppm is a bit less "niche" than pcm), and that'll still easily fit into the 32 bit integer return type.  Any thoughts?

dbrgn replied:
Ah, that makes sense. "per mille" is 1/1000 and "per cent mille" is one percent of a "per mille".

dbrgn reviewed (changes requested):
Sorry for the long delay! The issue fell off my todo list somehow.

I'd prefer newtypes that are named after their meaning, not after their internal units. Once that's fixed, I think this PR can be merged.

dbrgn reviewed (approved):
Thanks! It would be good to rustfmt the commits, but in the interest of getting this merged quickly, I'll do it myself 🙂

dbrgn commented:
Merged in 26868762269dc054eeba0c236352f9b38372b0f4, thanks!
